### ISSUE TITLE
Symfony 3.4.8: A Token was not found in the TokenStorage

### ISSUE THREAD

**Issue by fritzmg** [contao/core-bundle#1466]
Since the release of Symfony 3.4.8, the following error will occur in the front and back end of Contao 4.4.17 and 4.5.7:
```
Symfony\Component\Security\Core\Exception\AuthenticationCredentialsNotFoundException:
A Token was not found in the TokenStorage.

  at vendor\symfony\symfony\src\Symfony\Component\Security\Http\Firewall\AccessListener.php:51
  at Symfony\Component\Security\Http\Firewall\AccessListener->handle(object(GetResponseEvent))
     (vendor\symfony\symfony\src\Symfony\Bundle\SecurityBundle\Debug\WrappedListener.php:46)
  at Symfony\Bundle\SecurityBundle\Debug\WrappedListener->handle(object(GetResponseEvent))
     (vendor\symfony\symfony\src\Symfony\Bundle\SecurityBundle\Debug\TraceableFirewallListener.php:35)
  at Symfony\Bundle\SecurityBundle\Debug\TraceableFirewallListener->handleRequest(object(GetResponseEvent), object(RewindableGenerator))
     (vendor\symfony\symfony\src\Symfony\Component\Security\Http\Firewall.php:56)
  at Symfony\Component\Security\Http\Firewall->onKernelRequest(object(GetResponseEvent))
     (vendor\symfony\symfony\src\Symfony\Bundle\SecurityBundle\EventListener\FirewallListener.php:48)
  at Symfony\Bundle\SecurityBundle\EventListener\FirewallListener->onKernelRequest(object(GetResponseEvent), 'kernel.request', object(TraceableEventDispatcher))
  at call_user_func(array(object(TraceableFirewallListener), 'onKernelRequest'), object(GetResponseEvent), 'kernel.request', object(TraceableEventDispatcher))
     (vendor\symfony\symfony\src\Symfony\Component\EventDispatcher\Debug\WrappedListener.php:104)
  at Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke(object(GetResponseEvent), 'kernel.request', object(ContainerAwareEventDispatcher))
     (vendor\symfony\symfony\src\Symfony\Component\EventDispatcher\EventDispatcher.php:212)
  at Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(array(object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener)), 'kernel.request', object(GetResponseEvent))
     (vendor\symfony\symfony\src\Symfony\Component\EventDispatcher\EventDispatcher.php:44)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch('kernel.request', object(GetResponseEvent))
     (vendor\symfony\symfony\src\Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher.php:139)
  at Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch('kernel.request', object(GetResponseEvent))
     (vendor\symfony\symfony\src\Symfony\Component\HttpKernel\HttpKernel.php:127)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor\symfony\symfony\src\Symfony\Component\HttpKernel\HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor\symfony\symfony\src\Symfony\Component\HttpKernel\Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web\app_dev.php:64)
```

**Comment by christianbarkowsky** [comment]
I can confirm that.

**Comment by scuben** [comment]
Can confirm that too.

**Comment by leofeyer** [comment]
I cannot confirm this. My local installation works fine with Symfony 3.4.8.

**Comment by leofeyer** [comment]
@contao/developers Did you encounter this?

**Comment by Toflar** [comment]
Is that maybe due to old data in the session?

**Comment by fritzmg** [comment]
> Is that maybe due to old data in the session?

I have tried deleting all cookies and truncating `tl_session` in the affected installation. It had no effect. Is there something else I could try?

**Comment by aschempp** [comment]
I did not have this issue in any installation I updated (yet).

**Comment by adera-web** [comment]
I confirm.
Updateing from 4.4.16 to 4.4.17 I get:

[2018-04-10 16:39:20] app.CRITICAL: An exception occurred. {"exception":"[object] (Symfony\\Component\\Security\\Core\\Exception\\AuthenticationCredentialsNotFoundException(code: 0): A Token was not found in the TokenStorage. at /www/htdocs/w015b31c/c44/canvasplus/vendor/symfony/symfony/src/Symfony/Component/Security/Http/Firewall/AccessListener.php:51)"} []

**Comment by formatz** [comment]
I did reproduce the issue with Contao 4.4.17 and Symfony 3.4.8.

Changes made in class [SimpleAuthenticationProvider](https://github.com/symfony/symfony/blob/v3.4.8/src/Symfony/Component/Security/Core/Authentication/Provider/SimpleAuthenticationProvider.php#L54) cause a second call in UserProvider class with $user "anon." that trigger an AuthenticationException.
This exception clear then the TokenStorage value in [SimplePreAuthenticationListener](https://github.com/symfony/symfony/blob/v3.4.8/src/Symfony/Component/Security/Http/Firewall/SimplePreAuthenticationListener.php#L95).
AccessListener then detect a "null" token and rise the AuthenticationCredentialsNotFoundException. 

Is there a way to return a User object from [ContaoUserProvider::loadUserByUsername](https://github.com/contao/core-bundle/blob/4.4/src/Security/User/ContaoUserProvider.php#L61) where username "anon." is provided ?

**Comment by fritzmg** [comment]
@formatz thx for analysing! But do you have any idea why this does not happen in _all_ installations?

**Comment by formatz** [comment]
@fritzmg I tried with a fresh installation too - Symfony Standard Edition 3.4 + Contao 4.4 - and the error happened too.
You mentioned you tried a fresh install of Contao 4.4.7 and You was unable to reproduce the error. Was it with Contao Managed Edition ?

I didn't try with Contao 4.5 nor Contao Managed Edition

**Comment by fritzmg** [comment]
> You mentioned you tried a fresh install of Contao 4.4.7 and you was unable to reproduce the error. Was it with Contao Managed Edition ?

Yes, always with the Contao Managed Edition. The error occurs in one of my local, pre-existing Contao 4 (Managed Edition) installations as well after an update - I am not able to reproduce it with a __fresh__ Contao 4.4.17 (Managed Edition) installation though.

**Comment by pressi** [comment]
I had the same problem after an update from 4.4.8 to 4.4.17 (update with Contao Manager).
Online Server from All-inkl.

**Comment by aschempp** [comment]
I have found the problem for this: https://github.com/symfony/symfony/pull/26788#issuecomment-380586927

Due to the fact that the User in Contao is `BackendUser` or `FrontendUser`, the change in Symfony attempts to call the `UserProvider` with the instantiated user, which results in strange behavior because only a string is expected.

**Comment by leofeyer** [comment]
So I guess we have to add version `3.4.8` to the `symfony/security` conflict.

1. Does this affect Contao 4.4 only or also Contao 4.5?
2. Do we also have to add a conflict for Symfony `4.0.8`?

**Comment by aschempp** [comment]
As far as I can tell by the code, it should not affect 4.5, but people are reporting the same issue there :-(
A conflict for Symfony 4.0.8 should not be necessary as it's not a supported version (yet) anyway.

**Comment by fritzmg** [comment]
> As far as I can tell by the code, it should not affect 4.5, but people are reporting the same issue there :-(

I think I confused one installation where this happened with it being 4.5 when it actually was 4.4. I cannot reproduce it in any 4.5 installation currently.

**Comment by leofeyer** [comment]
We are no longer using the simple authentication provider in Contao 4.5, therefore it cannot be affected.

**Comment by aschempp** [comment]
Can someone please test the changes in https://github.com/contao/core-bundle/pull/1477 ?

**Comment by fritzmg** [comment]
The fix does not work unfortunately, the same error still happens.

I don't really understand the whole system, but here is what I see:

1. `$user = $authToken->getUser();` returns `'anon.'`.
2. `$user = $this->userProvider->loadUserByUsername($user);` throws a `UsernameNotFoundException`.

**Comment by formatz** [comment]
Same error again.

Should be possible to create a class AnonUser which extends \Contao\User and return an instance of this object from ContaoAuthenticator::authenticateToken() where the UsernameNotFoundException is catched and when $token->getSecret() === $providerKey ?
I should probably propose a pull request but is this an acceptable way to resolve the problem ?

**Comment by aschempp** [comment]
I don't think so. Symfony explicitly uses an `AnonynousToken` that does not have a user, that's a valid use case. I just reported it on the Symfony issue.

**Comment by aschempp** [comment]
That's the question of Symfony's implementation. You can't necessarily assume that `AnonymousToken` is the right thing to check. For example, there's a configuratio for the `AuthenticationTrustResolver` to set the specific class. Which does not mean that necessarily applies to us…
https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/Resources/config/security.xml#L8

**Comment by leofeyer** [comment]
I have extended the conflict with `symfony/security` in e55f3b6aaff0782c55175d5a07b90767586ea9c4 until the issues has been resolved.

**Comment by leofeyer** [comment]
If the changes from https://github.com/symfony/symfony/pull/27059 make their way into Symfony 3.4.9, we can change the conflict to `3.4.7 || 3.4.8`.

**Comment by Toflar** [comment]
Symfony's released new versions, @fritzmg do you mind testing them?

**Comment by fritzmg** [comment]
Contao `4.4.17` seems to work with Symfony `3.4.9` 👍 

I don't know how I could test Contao `4.4.18` with Symfony `3.4.9` though.

**Comment by fritzmg** [comment]
You need to add `3.3.17`, otherwise composer might switch back to that version where the same problem exists.

**Comment by leofeyer** [comment]
Thanks for the heads-up. Added in 7f6eb72a362ec3d9447f7ed70e7c1634c4cd333a.
